Upload (#1)
privacy-ethereum/nova-by-hand: part_2.md

The R1CS for $x^3 + x + 5 = y$ is structured as:

$$
\begin{bmatrix}
1 & 0 & 0 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 & 0 & 0 \\
1 & 0 & 1 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 0 & 5
\end{bmatrix} \cdot \begin{bmatrix}
\omega \\
\text{sym1} \\
y \\
\text{sym2} \\
\text{out} \\
1
\end{bmatrix} \odot \begin{bmatrix}
1 & 0 & 0 & 0 & 0 & 0 \\
1 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 & 1
\end{bmatrix} \cdot \begin{bmatrix}
\omega \\
\text{sym1} \\
y \\
\text{sym2} \\
\text{out} \\
1
\end{bmatrix} = \begin{bmatrix}
0 & 1 & 0 & 0 & 0 & 0 \\
0 & 0 & 1 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 0 & 0 \\
0 & 0 & 0 & 0 & 1 & 0
\end{bmatrix} \cdot \begin{bmatrix}
\omega \\
\text{sym1} \\
y \\
\text{sym2} \\
\text{out} \\
1
\end{bmatrix}
$$

In this statement, there are many instance vectors $z$ possible, as various pairs of $x, y$ satisfy $x^3 + x + 5 = y$.
For instance, when $(x, y) = (2, 15)$, $z$ is $[2, 4, 8, 10, 15, 1]$. Similarly, for $(x, y) = (3, 35)$, $z$ is $[3, 9, 27, 30, 35, 1]$.

## 2.3 R1CS in IVC

> Nova is one of the IVC, so we suppose that IVC has the function $F$ of the $x^3 + x + 5 = y$ and it's R1CS. This is the IVC, so the output $y$ at step $i$ becomes the input $x$ at the step $i+1$. 

Reviewer: ```suggestion
Suppose our IVC consists into repeatedly applying the function $F$ of the $x^3 + x + 5 = y$ through it's R1CS representation. In Nova's IVC, the output $y$ at step $i$ will become the input $x$ at step $i+1$. 
```

For instance,
Step 1: $(x, y) = (1, 7)$, $z$ is $[1, 1, 1, 2, 7, 1]$
Step 2: $(x, y) = (7, 355)$, $z$ is $[7, 49, 343, 350, 355, 1]$
Step 3: $(x,y)=(355,44739235)$, $z$ is $[355,126025,44738875,44739230,44739235,1]$.

This means the IVC maintains the same R1CS coefficient matrices $A, B, C$ throughout the entire steps, but there will be different instance vectors $z_i$.

Thus, the goal of Nova Folding is to fold two instance vectors $z_1, z_2$ into a single vector $z_{(1,2)}$. After IVC should fold the $z_{(1,2)}$ and $z_3$ into $z_{((1,2),3)}$, continue to fold following until the last.

![r1cs folding](/images/part2_r1csfolding.png)

This illustration is an intuitive representation of Folding. Let's look at the specifics. The actual Nova Folding does Folding of an Augmented Circuit.

> # 3 First Attempt: Try Folding to two R1CS

Reviewer: ```suggestion
# 3 First Attempt: Trying to naively fold two R1CS
```

\end{align}
$$

Then $Z = Z_1 + r \cdot Z_2$ becomes:

$$
\begin{align}
Z = [W_1 + r \cdot W_2, x_1 + r \cdot x_2, 1 + r \cdot 1]
\end{align}
$$

Now, expanding $A \cdot Z \odot B \cdot Z$ is below, but it is not equal to the $C \cdot Z=C \cdot (Z_1 + r \cdot Z_2)$

$$
\begin{align*}
A \cdot Z \odot B \cdot Z &= A \cdot (Z_1 + r \cdot Z_2) \odot B \cdot (Z_1 + r \cdot Z_2) \\
&= A \cdot Z_1 \odot B \cdot Z_1 + r \cdot (A \cdot Z_1 \odot B \cdot Z_2 + A \cdot Z_2 \odot B \cdot Z_1) + r^2 \cdot (A \cdot Z_2 \odot B \cdot Z_2) \\ 
&= C \cdot Z_1+r^2 (C \cdot Z_2)+r \cdot (A \cdot Z_1 \odot B \cdot Z_2 + A \cdot Z_2 \odot B \cdot Z_1) 
\\
&\neq C \cdot Z.
\end{align*}
$$

These are the causes.:
1. The cross term $r \cdot (A \cdot Z_1 \odot B \cdot Z_2 + A \cdot Z_2 \odot B \cdot Z_1)$ appears unexpectedly, complicating the equation.
2. The presence of the $r^2$ term also adds complexity not accounted for in the original $C \cdot Z$ equation.
3. The scalar multiplication with $r$ alters the constant term, $1$, into $r$ in the vector $Z$, $Z = [W_1 + r \cdot W_2, x_1 + r \cdot x_2, 1 + r \cdot 1]$ and deviating from the standard format expected in a valid R1CS instance.

> ## 3.2 How can we solve this R1CS isuue?

Reviewer: ```suggestion
## 3.2 How can we solve this R1CS issue?
```

To overcome the challenges from the first try, we have to adopt a new strategy in the second attempt by introducing two key elements:

- Introduction of an Error Vector $E$: Introduce an error vector $E$ to absorb the cross terms, $r \cdot (A \cdot Z_1 \odot B \cdot Z_2 + A \cdot Z_2 \odot B \cdot Z_1)$ , from folding $Z_1$ and $Z_2$.

- Introduction of a Scalar $u$: To address the issues with the $r^2$ term and the altered constant in $Z$, we introduce a scalar $u$. This scalar adjusts for the factor of $r$ in $CZ_1 + r^2 \cdot CZ_2$ and the modified constant term in $Z = (W, x, 1 + r \cdot 1)$.

This approach leads to a variant of R1CS, termed Relaxed R1CS.

# 4 Relaxed R1CS

## 4.1 Relaxed R1CS
### 4.1.1 Definition of Relaxed R1CS

In Relaxed R1CS, the original R1CS is  adjusted to accommodate the additional terms introduced by $E$ and $u$. The Relaxed R1CS can be expressed as follows:

$$ (A \cdot Z) \odot (B \cdot Z) = u \cdot (C\cdot Z) + E$$

Here, $E$ serves to relax the cross-term effects. The scalar $u$ is introduced to adjust for the $r^2$ term and the altered constant in $Z$, ensuring the equation's balance and alignment.

> This modification allows the linear-combinationed vector $Z$, resulting from the folding process, to still satisfy the Relaxed R1CS. Note that any instance of a Relaxed R1CS can be resented as R1CS by adopting $u = 1$ and $E = 0$. So Relaxed R1CS is still  the property of NP-completeness.

Reviewer: ```suggestion
This modification allows the linear-combinationed vector $Z$, resulting from the folding process, to still satisfy the Relaxed R1CS. Note that any instance of a Relaxed R1CS can be represented using R1CS by adopting $u = 1$ and $E = 0$. So Relaxed R1CS retains the property of NP-completeness.
```

x &= x_1 + r \cdot x_2 \\
W &= W_1 + r \cdot W_2 \\
\end{align*}
$$

Consequently, $Z = (W, \mathsf{x}, u)$ becomes a combination of $Z_1$ and $Z_2$:

$$\begin{align*}
AZ \odot BZ &= A(Z_1 + r\cdot Z_2) \odot B(Z_1 + r\cdot Z_2) \\
&= AZ_1 \odot BZ_1 + r \cdot(AZ_1 \odot BZ_2 + AZ_2 \odot BZ_1) + r^2 \cdot (AZ_2 \odot BZ_2 )\\
&= u_1\cdot CZ_1 + E_1 + r \cdot (AZ_1 \odot BZ_2 + AZ_2 \odot BZ_1) + r^2 \cdot (u_2 \cdot CZ_2 + E_2) \\
&= u_1 \cdot CZ_1 + r \cdot u_2 \cdot CZ_1 - r \cdot u_2\cdot CZ_1 + E_1 + r \cdot (AZ_1 \odot BZ_2 + AZ_2 \odot BZ_1) + r^2 \cdot (u_2 \cdot CZ_2 + E_2) + r \cdot u_1 \cdot CZ_2 - r \cdot u_1 \cdot CZ_2 \\
&= (u_1 + r \cdot u_2) \cdot CZ_1 + E_1 -r \cdot u_2 \cdot CZ_1 + r \cdot (AZ_1 \odot BZ_2 + AZ_2 \odot BZ_1) + (u_1+ r \cdot u_2) \cdot r \cdot CZ_2 + r^2 \cdot E_2 - r \cdot u_1 \cdot CZ_2 \\
&=(u_1 + r \cdot u_2) \cdot C (Z_1 + rZ_2) + r \cdot (AZ_1 \odot BZ_2 + AZ_2 \odot BZ_1 - u_2 \cdot CZ_1 -u_1 \cdot CZ_2) + E_1 + r^2 \cdot E_2 \\
&= u \cdot (CZ) + E
\end{align*}$$

## 4.2 Prover and Verifier Actions in Relaxed R1CS

This section describes how Prover and Verifier interact to build the Relaxed R1CS, although this is a naïve approach.

> ### 4.2.0 What do they have by begin?

Reviewer: ```suggestion
### 4.2.0 What do they have at the beginning of the protocol?
```

![4.2.0](/images/part2_4.2.0.png)

> At first, the Prover has an instance-witness pair for the Relaxed R1CS, denoted as $((E_i,u_i,x_i),W_i)$, and the Verifier knows the only instance $(E_i,u_i,x_i)$ for it.

Reviewer: ```suggestion
At first, the Prover has two instance-witness pairs satisfying a Relaxed R1CS equation. Each are denoted as $((E_i,u_i,x_i),W_i)$, and the Verifier knows the only instance $(E_i,u_i,x_i)$ for it.
```

- Non-Triviality: In the initial scheme, the prover sends witnesses $(W_1, W_2)$ for the verifier to compute $E$, $(E = E_1 + r \cdot (A Z_1 \odot B Z_2 + A Z_2 \odot B Z_1 - u_1 \cdot C Z_2 - u_2 \cdot C Z_1) + r^2 \cdot E_2)$, which can be very large. This process makes the folding scheme not trivial.

- Lack of Zero-Knowledge: The scheme also fails to provide zero-knowledge properties. Since the witnesses are sent directly to verifier, it compromises the privacy.

### 4.3.2 Solution
To address these issues, the final protocol introduces a new concept: Committed Relaxed R1CS. This variant involves key modifications:

- Use of Homomorphic Commitments: The protocol employs additively homomorphic commitments to both the witness $W$ and the slack vector $E$.

- Treating $W$ and $E$ as the Witness: Both $W$ and $E$ are treated as parts of the witness.

- Single Commitment from the Prover: Instead of sending raw witnesses themselves, the prover sends commitment of the witness. This commitment aids the verifier in computing commitments to the folded witness $(W, E)$ efficiently.

# 5 Committed Relaxed R1CS

> Before moving on to the Committed Relaxed R1CS, let's check the what commitemnt scheme is.

Reviewer: ```suggestion
Before moving on to the Committed Relaxed R1CS, let's check what a commitment scheme is.
```

### 5.1.2 Additively Homomorphic Property
Additive homomorphism is a property where arithmetic operations (in this case, addition) performed on encrypted data correspond to the same operations on the original data.

In the Pedersen Commitment scheme:
1. For two values $x$ and $y$, their commitments $c(x, r) = xP + rQ$ and $c(y, s) = yP + sQ$ are created.
2. Adding these commitments results in $c(x, r) + c(y, s) = (x + y)P + (r + s)Q$, forming a new commitment for $x + y$.

## 5.2 Definition of Committed Relaxed R1CS
![5.2](/images/part2_5.2.png)

In the Relaxed R1CS, instance-witness pairs are denoted as $(E_i, u_i, W_i, x_i)$. In terms of the Committed Relaxed R1CS, the witness of the Committed Relaxed R1CS can be defined as $(E_i, r_{E_i}, W_i, r_{W_i})$ and the instance of the Committed Relaxed R1CS can be defined as $(\bar{E}_i, u_i, \bar{W}_i, x_i)$. Here, $r$ is a random number.

# 6 Folding Scheme for Committed Relaxed R1CS

Let's see how to construct Folding scheme for Committed Relaxed R1CS.

## 6.1 Prover and Verifier Actions in the Folding Scheme for Committed Relaxed R1CS

> ### 6.1.0 What do they have by begin?

Reviewer: ```suggestion
### 6.1.0 What do they have at the beginning of the protocol?
```

There are two primary functions of the augmented function $F'$:

1. Incremental Computation: This involves the execution of function $F$, which takes $u_i$ as its input, containing $z_i$, and outputs $Z_{i+1} = F(Z_i)$.

2. Folding Verifier: The NIFS.P folds from $U_{\text{acc}, i}$ and $u_i$ to $U_{\text{acc}, i+1},$. So $F'$ calls the $NIFS.V$ to check it.

Additionally, the Augmented Function $F'$ incorporates Augmented Constraints to generate the IVC Proof.

![augmented instance](/images/part2_augmentedconstraintsinstance.png)

And in IVC, the Prover computes a new instance of $u_{i+1}$, indicating that the computation of $i+1$ in $F'$ has been executed correctly. The component $u_{i+1}.x$ is instantiated by a hash function whose inputs are $i+1$, $Z_0$, $Z_{i+1}$, and $U_{\text{acc}, i+1}$, ensuring it becomes a constant value.

![augmented pic](/images/part2_augmentedconstraintspic.png)

> This illustration is a mixture of Figure 4 on P.18 of the Nova paper and [Revisiting the Nova Proof System - Wilson Nguyen](https://youtu.be/h_PU7FZWiQk?si=G5DZYfrlQQi8SG-j).

Reviewer: I think you can also add a link to the nova pdf directly
